test/src/utils/Transactions.test.js

```diff
@@ -0,0 +1,132 @@
+      assertGasLt(tx, 1000000);
+    });
+
+    it('uses specified gas', async function () {
+      const { tx } = await sendTransaction(this.instance.initialize, [42, 'foo', [1,2,3]], { gas: 800000});
+      assertGas(tx, 800000);
+    });
+
+    it('honours other tx params', async function () {
+      const { tx } = await sendTransaction(this.instance.initialize, [42, 'foo', [1,2,3]], { from: account2 });
+      assertGasLt(tx, 1000000);
+      assertFrom(tx, account2);
+    });
+
+    it('handles failing transactions', async function () {
+      await assertRevert(sendTransaction(this.instance.reverts));
+    });
+  });
+
+  describe('deploy', function () {
+    describe('without a constructor', function () {
+      it('correctly deploys an instance', async function () {
+        const instance = await deploy(this.DummyImplementation);
+        (await instance.version()).should.eq("V1");
+      });
+  
+      it('estimates gas', async function () {
+        const instance = await deploy(this.DummyImplementation);
+        assertGasLt(instance.transactionHash, 1000000);
+      });
+  
+      it('uses specified gas', async function () {
+        const instance = await deploy(this.DummyImplementation, [], { gas: 800000 });
+        assertGas(instance.transactionHash, 800000);
+      });
+  
+      it('honours other tx params', async function () {
+        const instance = await deploy(this.DummyImplementation, [], { from: account2 });
+        assertGasLt(instance.transactionHash, 1000000);
+        assertFrom(instance.transactionHash, account2);
+      });
+    });
+
+    describe('with a constructor', function () {
+      beforeEach('load contract', function () {
+        this.WithConstructorImplementation = Contracts.getFromLocal('WithConstructorImplementation');
+      });
+
+      it('correctly deploys an instance', async function () {
+        const instance = await deploy(this.WithConstructorImplementation, [42, "foo"]);
+        (await instance.value()).toNumber().should.eq(42);
+        (await instance.text()).should.eq("foo");
+      });
+
+      it('estimates gas', async function () {
+        const instance = await deploy(this.WithConstructorImplementation, [42, "foo"]);
+        assertGasLt(instance.transactionHash, 1000000);
+      });
+
+      it('uses specified gas', async function () {
+        const instance = await deploy(this.WithConstructorImplementation, [42, "foo"], { gas: 800000 });
+        assertGas(instance.transactionHash, 800000);
+      });
+  
+      it('honours other tx params', async function () {
+        const instance = await deploy(this.WithConstructorImplementation, [42, "foo"], { from: account2 });
+        assertGasLt(instance.transactionHash, 1000000);
+        assertFrom(instance.transactionHash, account2);
+      });
+
+      it('handles failing constructors', async function () {
+        await assertRevert(deploy(this.WithConstructorImplementation, [0, "foo"]));
```

Comment:
My previous question rather was about how `estimateGas` behaves if the underlying transaction reverts. I suppose it estimates how much gas is needed to hit the `revert`?

I have a feeling that I've seen some tools (MyCrypto?) warn me before sending a transaction that is guaranteed to revert, and gas estimation sounds like the moment when they can figure that out. Maybe we should just open another issue to consider implementing such warnings. WDYT?

Reply:
Actually `estimateGas` returns an error if the transaction fails. On geth I'm getting a `Error: gas required exceeds allowance or always failing transaction`. This means that we will never send the tx, but actually fail beforehand. So this optimization is already in place :-)